instance/gandi.py

```diff
@@ -189,4 +192,157 @@ def remove_dns_record_callback(zone_id, zone_version):
+
+        if not self._domain_cache:
+            logger.warning(
+                'Using the default instance base domain %s as the fallback. '
+                'This will not work if the API key has no permission to manage the DNS records for this domain',
+                settings.DEFAULT_INSTANCE_BASE_DOMAIN
+            )
+            self._domain_cache = [settings.DEFAULT_INSTANCE_BASE_DOMAIN]
+
+    def _split_domain_name(self, domain):
+        """
+        Split the given domain name to the registered domain and the subdomain.
+        """
+        if self._domain_cache is None:
+            self._populate_domain_cache()
+        labels = domain.lower().split('.')
+        for split_index in range(len(labels) - 1):
+            registered_domain = '.'.join(labels[split_index:])
+            if registered_domain in self._domain_cache:
+                subdomain = '.'.join(labels[:split_index]) or '@'
+                return subdomain, registered_domain
+        raise ValueError(
+            'The given domain name "{}" does not match any domain registered in the Gandi account.'.format(domain)
+        )
+
+    def _get_base_config(self):
+        """
+        Return the common base configuration for all DNS operation API calls.
+        """
+        return {
+            "provider_name": "gandi",
+            "gandi": {
+                "auth_token": self.api_key,
+                "api_protocol": "rest"
+            }
+        }
+
+    def _dns_operation(self, callback, log_msg, attempts=4, retry_delay=1):
+        """
+        Encapsulate logic that is common to high-level DNS operations: grab the global lock, do the operation,
+        and retry the whole procedure multiple times if necessary.
+        """
+        with cache.lock('gandi_set_dns_record'):  # Only do one DNS update at a time
+            for i in range(1, attempts + 1):
+                try:
+                    logger.info('%s (attempt %d out %d)', log_msg, i, attempts)
+                    result = callback()
+                    break
+                except Exception:  # pylint: disable=broad-except
+                    if i == attempts:
+                        raise
+                    time.sleep(retry_delay)
+                    retry_delay *= 2
+        return result
+
+    def add_dns_record(self, record):
+        """
+        Add a DNS record to the domain.
+        """
+        lexicon_config = self._get_base_config()
+        lexicon_config['domain'] = record['domain']
+        lexicon_config['action'] = 'create'
+        lexicon_config['type'] = record['type']
+        lexicon_config['name'] = record['name']
+        lexicon_config['content'] = record['value']
+        lexicon_config['ttl'] = record['ttl']
+        config = ConfigResolver()
+        config.with_dict(dict_object=lexicon_config)
+        client = Client(config)
+        result = client.execute()
+        return result
```

Comment:
Can we do some basic error handling internally? 
We should handle the result internally and throw a known error when an issue happens. 
Example: 
- **If adding a DNS record succeeds:** no need to return anything
- **If it fails:** return a custom exception class with error (`SetDnsFailed` or something like that that that allows error handling and logging on modules using this class).

Reply:
@giovannicimolin, the lexicon code only returns an `Exception` instance with a custom error message. The only improvement that can be done here is to catch all exceptions, log an error message and re-raise the exception since we cannot proceed further without setting the DNS record. There is no error handling in the XML-RPC API client code as well.